Implement CD Audio using Linux ioctls

weirddan455 opened the pull request:
Fixes #2664

I've mostly de-coupled the Linux ioctl interface from SDL.  The only thing it still uses SDL for is getting device names (the UI commands use SDL device numbers so they need to match up).  This removed a kind of messy inheritance thing.

Most of the CD interface functions DOSBox uses mapped pretty cleanly to the Linux ioctls.  The hardest part was getting the mixer stuff right.  I've tested and it sounds correct to me now though.

I also tested that this works with CDEmu (https://cdemu.sourceforge.io/).  That creates a virtual CD drive and can maybe be useful if someone has an image that DOSBox won't mount.

Tested with Secret of Monkey Island (image through CDEmu) and Quake (physical CD).

johnnovak commented on src/dos/cdrom.h:
```diff
@@ -354,18 +349,42 @@ class CDROM_Interface_Image final : public CDROM_Interface
+
+	int cdrom_fd                            = -1;
+	mixer_channel_t mixer_channel           = nullptr;
+	std::unique_ptr<int16_t[]> audio_buffer = nullptr;
```
It would be better to use unsigned unless we really need signed, so `uint16_t`.

weirddan455 replied:
Well, the data is actually 16 bit signed integers because that's how CD audio is stored.  As of now, I'm not directly manipulating it so it could technically be anything (though it does need to be explicitly cast to `int16_t` before passing the pointer off to `AddSamples_s16`)

johnnovak replied:
Ah right, then it makes sense to use `int16_t.

johnnovak commented on src/dos/cdrom.h:
```diff
@@ -354,18 +349,42 @@ class CDROM_Interface_Image final : public CDROM_Interface
+	size_t audio_position                   = 0;
+	size_t frames_in_buffer                 = 0;
+	int current_sector                      = 0;
+	int sectors_remaining                   = 0;
```
`uint16_t` for both (or the smallest integer we can use).

weirddan455 replied:
It definitely can't be `uint16_t` since we have this constant `#define MAX_REDBOOK_SECTOR 1826090u`.  Could change it to be unsigned since negative values don't make sense.  I just left it `int` because that's ultimately what the `ioctl` returns and takes for sector values.

johnnovak replied:
Yeah, we try to do better than the loose C way of doing things, but probably ok for this specific case.

johnnovak commented on src/dos/cdrom.h:
```diff
@@ -354,18 +349,42 @@ class CDROM_Interface_Image final : public CDROM_Interface
+	int current_sector                      = 0;
+	int sectors_remaining                   = 0;
+	bool playing                            = false;
+	bool paused                             = false;
```
minor: is_playing and is_paused?

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+	         leadOut.fr);
+#endif
+	return true;
+}
```
A few blank lines here and there would help readability; now this function is just one big block of text.

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom.h:
```diff
@@ -354,18 +349,42 @@ class CDROM_Interface_Image final : public CDROM_Interface
+	void CDAudioCallback(uint16_t requested_frames);
+	void InitAudio();
+
+	int cdrom_fd                            = -1;
```
Can be unsigned?

weirddan455 replied:
Nope.  Linux file descriptors explicitly return -1 for error so I'm using that as the default "invalid/initialize" value.

johnnovak replied:
Ok, it's been a while I used C 😄

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+bool CDROM_Interface_Ioctl::GetAudioTrackInfo(uint8_t track, TMSF& start,
+                                              unsigned char& attr)
+{
+	if (cdrom_fd == -1) {
```
It would be better to have an `is_opened` flag instead of deriving that information from magic values in multiple places. Then this becomes `if (is_opened)`.

weirddan455 replied:
Could do that I guess.  I think it just makes things more complex.  When you do an `open`, it's always -1 if error.  Anything else is valid.  0 is even valid (it's stdin).

I'd rather do a macro or something if readability is a concern.  Those checks can maybe even be omitted entirely since the `ioctl` will fail and get handled later anyway (though it does avoid potentially expensive syscalls if we know they're going to fail ahead of time).

johnnovak replied:
> I wrote an `IsOpen()` function that checks whether the file descriptor is in error state. I'm a little afraid if I make a separate boolean that there could be bugs if one of the variables gets changed but not the other.

It's fine, it's not that complex. We do similar things in a number of places.

> Hopefully this improves readability while keeping the data in one place.

Yes, I think it does. Reduced duplication and gets rid of "deriving" the state from magic values in file descriptors, which although not a huge deal, it's not ideal either.

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+	         msf_to_frames(relPos));
+#endif
+	return true;
+}
```
Same general comment: please insert some blank lines to group things logically and break up the wall-of-text look. This is hard to parse visually. I'd insert blanks around the `if` and before the last debug log statement.

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+	mediaPresent = false;
+	mediaChanged = false;
+	trayOpen     = false;
+	if (cdrom_fd == -1) {
```
Blank above.

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+	case CDS_TRAY_OPEN: trayOpen = true; break;
+	case CDS_DISC_OK: mediaPresent = true; break;
+	}
+	int ret = ioctl(cdrom_fd, CDROM_MEDIA_CHANGED, CDSL_CURRENT);
```
Blank above.

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+#endif
+		return true;
+	}
+	switch (ioctl(cdrom_fd, CDROM_DRIVE_STATUS, CDSL_CURRENT)) {
```
Blank above.

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+	}
+	int ret = ioctl(cdrom_fd, CDROM_MEDIA_CHANGED, CDSL_CURRENT);
+	mediaChanged = (ret > 0) && (ret & 1);
+#ifdef DEBUG_IOCTL
```
Blank above.
(Stopping now... you get the picture 😄 )

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+	         mediaPresent ? "present" : "not present",
+	         mediaChanged ? "was changed" : "hasn't been changed",
+	         trayOpen ? "open" : "closed");
+#endif
```
Very nice debug logging 💯

weirddan455 replied:
Yeah, I stole those from `cdrom_image.cpp` :smile:

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -92,17 +252,35 @@ bool CDROM_Interface_Ioctl::ReadSectors(PhysPt buffer,
 	}
 
-	return success;
+	return false;
```
Ok, so the vertical spacing/grouping is nicely done in this function 🎉  I'd expect the same in all the others.

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +291,154 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+	// unique or they get overwritten)
+	mixer_channel_t exists;
+	int num                  = 0;
+	const std::string prefix = MixerChannelNamePrefix;
```
Why not just use `MixerChannelNamePrefix` directly below?

weirddan455 replied:
Because it needs to be converted to `std::string`

johnnovak replied:
Ah ok.

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +291,154 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+	// Find a a channel name that doesn't exist (channel names must be
+	// unique or they get overwritten)
+	mixer_channel_t exists;
+	int num                  = 0;
```
`auto`

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -92,17 +252,35 @@ bool CDROM_Interface_Ioctl::ReadSectors(PhysPt buffer,
-		else
-			success = false;
+	for (int i = 0; i < num; i++) {
+		const char* cd_name = SDL_CDName(i);
```
Use auto?

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -92,17 +252,35 @@ bool CDROM_Interface_Ioctl::ReadSectors(PhysPt buffer,
+
+	int num = SDL_CDNumDrives();
+	if ((cd_number >= 0) && (cd_number < num)) {
+		const char* cd_name = SDL_CDName(cd_number);
```
Use auto?

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+	case CDS_TRAY_OPEN: trayOpen = true; break;
+	case CDS_DISC_OK: mediaPresent = true; break;
+	}
+	int ret = ioctl(cdrom_fd, CDROM_MEDIA_CHANGED, CDSL_CURRENT);
```
Use auto?

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -92,17 +252,35 @@ bool CDROM_Interface_Ioctl::ReadSectors(PhysPt buffer,
+
+	if (cdrom_fd != -1) {
+		close(cdrom_fd);
+		cdrom_fd = -1;
```
Yeah, so please introduce that `is_opened` flag.

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +291,154 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+	} while (exists);
+
+	audio_buffer = std::make_unique<int16_t[]>(AudioBufferBytes / 2);
+	mixer_channel = MIXER_AddChannel(std::bind(&CDROM_Interface_Ioctl::CDAudioCallback,
```
Maybe break the `std::bind` stuff out into a temp var.

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +291,154 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+			LOG_WARNING("CDROM_IOCTL: CDAudioCallback: CDROMREADAUDIO ioctl failed");
+#endif
+		}
+	}
```
Vertical spacing please...

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom.h:
```diff
@@ -354,18 +349,42 @@ class CDROM_Interface_Image final : public CDROM_Interface
+
+	int cdrom_fd                            = -1;
+	mixer_channel_t mixer_channel           = nullptr;
+	std::unique_ptr<int16_t[]> audio_buffer = nullptr;
```
It would be better to use `std::vector`.

weirddan455 replied:
I was considering that.  I'm never using any vector operations though and it's always a fixed size.  The buffer is only used to read in one second's worth of data from a read `ioctl` and then pass it on (a millisecond or so at a time) to `AddSamples_s16`.  Both of which take raw pointers.

I'll change it if you want.  I don't have strong feeling about it.  But all I would be doing is calling `.data()` on it to get to the raw buffer so I'm not really sure what the benefit is.

johnnovak replied:
Yeah, fair points but vector is just nicer and safer, and that would remove the need for the unique_ptr cruft and memset. Not that important for this specific use case, but it contributes to the overall consistency of the codebase. And in most cases there is no reason to really _prefer_ raw arrays over vector or even std:array.

weirddan455 replied:
I changed this to use `std::vector`

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +291,154 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+		// Maximum this ioctl can read in a single call is 1 second (75
+		// redbook frames)
+		cd.nframes = std::min(sectors_remaining, CD_FRAMES);
+		constexpr int PCMFramesPerRedbookFrame = BYTES_PER_RAW_REDBOOK_FRAME /
```
Minor: Pcm instead of PCM (only capitalise the first letter of acronyms... yes, I know it's all over the place in the existing code...)

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +291,154 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+		frames_in_buffer = cd.nframes * PCMFramesPerRedbookFrame;
+		current_sector += cd.nframes;
+		sectors_remaining -= cd.nframes;
+		if (cdrom_fd == -1 || ioctl(cdrom_fd, CDROMREADAUDIO, &cd) != 0) {
```
Yeah, so this is too much... `is_opened` would be much better (or `is_open`).

johnnovak replied:
I wonder why you're not checking for `is_open` at the top of the function though?

weirddan455 replied:
lol ok i'm seeing your point on this

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +291,154 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+		sectors_remaining -= cd.nframes;
+		if (cdrom_fd == -1 || ioctl(cdrom_fd, CDROMREADAUDIO, &cd) != 0) {
+			// Write out silence if something goes wrong
+			memset(audio_buffer.get(), 0, AudioBufferBytes);
```
Use vector then you can use `std::fill` 😄

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +291,154 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+
+	// audio_position must be doubled since a frame is two int16_t's for stereo
+	// sound Does this need to be AddSamples_s16_nonnative on big endian?
+	mixer_channel->AddSamples_s16(frames_to_write,
```
Ideally, we'd want a full float mixer path; see this ticket for further details:
https://github.com/dosbox-staging/dosbox-staging/issues/1720

But this is kind of a special case as you can only retrieve the raw CD Audio data as a stream of integers. Anyway, just something to keep in mind for the future.

johnnovak replied:
Nice one 🎉

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +291,154 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+	                                        frames_available);
+
+	// audio_position must be doubled since a frame is two int16_t's for stereo
+	// sound Does this need to be AddSamples_s16_nonnative on big endian?
```
Yes, you need `AddSamples_s16_nonnative`. Just do the same thing as `sblaster.cpp` for now.

weirddan455 replied:
Fixed

johnnovak commented on src/dos/program_mount.cpp:
```diff
@@ -462,20 +462,21 @@ void MOUNT::AddMessages() {
 	        "\n"
 	        "Notes:\n"
 	        "  - '-t overlay' redirects writes for mounted drive to another directory.\n"
+	        "  - '-usecd #' gives direct access to a CD-ROM drive (needed for CD audio).\n"
```
`-usecd ID` would be better, then explain in a few words what ID is supposed to be.

weirddan455 replied:
Is it broken on Mac?  I don't have one.  I do dual boot Windows but haven't tested there.  Anyway, that section is already `#ifdef`'d for platform specific notes so I can put a note about that.

johnnovak replied:
Based on what others have reported (I never tried it), it only works on Windows, so now we have Windows support but via the SDL mixer, not the DOSBox mixer, and with this feature Linux support via the DOSBox mixer, and that's it. No support on macOS.

weirddan455 replied:
From what I've seen, SDL CD support doesn't use **any** mixer.  It doesn't actually read digital audio.  It just issues a command to the CD drive to tell it to start playing.  On Linux, this method is just broken on modern systems which is why I wrote this.  Maybe this method still works on Windows (this is how DOS games worked anyway) but there's no sound code here.  If anything gets mixed, it's up to the OS.

SDL does have code for Mac OSX, though no idea how well it works.  I'll test on Windows later.

johnnovak replied:
>  It just issues a command to the CD drive to tell it to start playing.

Right, it's even worse then than I thought 😅 

> SDL does have code for Mac OSX, though no idea how well it works. I'll test on Windows later.

Thanks for that. I have a rather low interest in this feature, and the guy who (re-)introduced it disappeared 🤷🏻

weirddan455 replied:
Yeah, that SDL code is pretty janky.  It's ripped out of version 1 which is just unmaintained nowadays.  If I was doing it, I would have at least tried to use the sdl12-compat library (re-implementation of the SDL 1 API using SDL 2 as a backend):  https://github.com/libsdl-org/sdl12-compat

That one looks like it tries to do digital audio properly.  I also have a low interest in messing with library code for other platforms though.

johnnovak replied:
Fair enough. I wasn't involved, but I guess nobody was super intersted in this and went "yeah, whatever" 🤷🏻

johnnovak replied:
So can you change `-usecd` to this (just format it properly, and feel free to reword it better and/or more succinctly):
```
'-usecd ID' gives direct access to a CD-ROM drive. This is needed for CD audio (only supported on Windows and Linux). Run 'mount -cd' to find out the list of valid IDs.
```

weirddan455 replied:
I changed it to the requested text.  Not sure how best to format it.  You gave me 3 lines of text there and now the help command is paging.

![mount](https://github.com/dosbox-staging/dosbox-staging/assets/8128047/cc32866b-16de-4362-8e0d-2dfd090bd457)

johnnovak commented:
Great work @weirddan455 🎉 

An overarching question/comment: it seems you're creating `CDROM_IOCTL_0`, `CDROM_IOCTL_1`, etc. mixer channels for every single mounted physical CD-ROM drive. That's problematic for a couple of reasons:

- Currently, we use the `CDAUDIO` mixer channel for mounted images that have CD audio, and I'm pretty sure it's the same deal with [FLAC/WAV/MP3/Opus tracks](https://dosbox-staging.github.io/downloads/release-notes/0.75.0/#support-flac-opus-and-mp3-cd-da-tracks) (but I've never personally used this feature).
- So at the very least the channel names should start with `CDAUDIO`, not `CDROM`. `IOCTL` seems also very OS specific; we don't want that. What if next month someone will add similar mixer-level CD Audio support for Windows? We don't want `CDAUDIO_IOCTL_0` on Linux, `CDAUDIO_WINDOWS_0`, then `CDAUDIO_MAC_0` or something some day... At the very least, I suggest using the names `CDAUDIO_0`, `CDAUDIO_1`, and so on. And then `CDAUDIO` is always dedicated to the mixer channel used by mounted CD images.
- But then, I'm not 100% sure it's a good thing to suddenly have 2-3-4+ CDAUDIO channels... In actual practice, how likely is it that CD Audio would be played simultaneously from multiple mounted CDs, and maybe from a mounted image too? Probably zero? I don't think this would ever happen in real life. Maybe technically possible, but it's more of a curiosity than something people would actually do. Therefore, unless you or someone else can convince me with some very strong arguments, I think we should just mix everything into the single `CDAUDIO` channel and job done. In 99.9999% of the cases (and I would place a good bet on this), only a single CD drive would stream its audio output into this channel. But in the extreme edge case scenario when we'd end up with output from multiple drives being mixed together, the user can simply just lower the volume of the `CDAUDIO` channel to avoid clipping (and yes, we can do without clipping because of the float mixing path!).

Okay, so let's just do that: single `CDAUDIO` mixer channel, and every kind of CD Audio output gets mixed into that. You can even test that when mounting a physical CD and a BIN/CUE image with audio tracks (well, maybe not the mixing of the two sources together as that would be pretty hard to trigger, but just that the audio gets mixed into the same mixer channel when both are mounted).

weirddan455 commented:
I don't think I can share a single mixer channel with the CDROM image class.  Whenever you create a mixer, it gets tied to a specific callback function (and if you use `std::bind` like I did and the image class does, it gets tied to a specific object as well).

I agree it's not ideal to be creating multiple channels.  I am creating mine lazily at least so the channel only gets created when music actually starts playing, rather when the device gets mounted.

johnnovak commented:
> I agree it's not ideal to be creating multiple channels. I am creating mine lazily at least so the channel only gets created when music actually starts playing, rather when the device gets mounted.

Yep, that's no good either because then the user cannot adjust the volume level of the channel before launching the game. It should be done when the drive is mounted.

Okay, I will look into sharing the same mixer channel across multiple devices. We should do this once and do it well.

weirddan455 commented:
>  Yep, that's no good either because then the user cannot adjust the volume level of the CD Audio channel before launching the game. It should be done when the drive is mounted.

How would the user adjust the volume before loading the game?  I didn't know that was a feature.  Quake has a volume slider in-game and I tested that worked.

johnnovak commented:
> > Yep, that's no good either because then the user cannot adjust the volume level of the CD Audio channel before launching the game. It should be done when the drive is mounted.
> 
> How would the user adjust the volume before loading the game? I didn't know that was a feature. Quake has a volume slider in-game and I tested that worked.

Oh man, you're in for a treat! 😎 

<img width="1215" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/1b131549-a447-4642-8a5f-f451752d6e1f">

<img width="1215" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/6374ab6c-2534-4b73-90a3-3522b0ce6b44">

Make sure to read these too (but you can read the whole audio section in the release notes; it's full of useful info):
https://dosbox-staging.github.io/downloads/release-notes/0.79.0/#general-mixer-improvements
https://dosbox-staging.github.io/downloads/release-notes/0.79.0/#reverb-and-chorus
https://dosbox-staging.github.io/downloads/release-notes/0.79.0/#crossfeed

A bunch of game specific audio configs by me:
https://github.com/dosbox-staging/dosbox-staging/wiki/Audio-configuration-recommendations

Mixer diagram (I need to update it with some more recent additions):
https://github.com/dosbox-staging/dosbox-staging/wiki/Audio-mixer-signal-flow-diagram

johnnovak commented:
Okay, I've done some digging and indeed, it would complicate things a lot to share mixer channels. `CDAUDIO_#` will do, where `#` is the ID of the device. That should be reasonably cross-platform too, should we add mixer-level CD Audio support on Windows and macOS in the future.

At the end of the day, this is a very niche feature that will eventually die out (as will the lasers in all CD-ROMs in existence in the not so distant future), so I don't want to make you do a lot of unnecessary work.

But the creation of the mixer channels must be done at the time the drive gets mounted.

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -347,9 +347,13 @@ void CDROM_Interface_Ioctl::CDAudioCallback(uint16_t requested_frames)
 	                                        frames_available);
 
 	// audio_position must be doubled since a frame is two int16_t's for stereo
-	// sound Does this need to be AddSamples_s16_nonnative on big endian?
```
minor: add back "sound" at the end of the sentence 😄

weirddan455 commented:
> Okay, I've done some digging and indeed, it would complicate things a lot to share mixer channels. `CDAUDIO_#` will do, where `#` is the ID of the device. That should be reasonably cross-platform too, should we add mixer-level CD Audio support on Windows and macOS in the future.
> 
> At the end of the day, this is a very niche feature that will eventually die out (as will the lasers in all CD-ROMs in existence in the not so distant future), so I don't want to make you do a lot of unnecessary work.
> 
> But the creation of the mixer channels must be done at the time the drive gets mounted.

@johnnovak I think I've addressed all your concerns now.  I changed this to do as you said.  Good idea on using device ID for the name.  That gets rid of the loop checking for available names since we can just assume only 1 channel per device.

Thanks again for the fast review :smile:

FeralChild64 commented:
I confirm that it works - on my at least the DOS Navigator can now play audio CDs on my Linux system :)

johnnovak commented on src/dos/cdrom.h:
```diff
@@ -354,18 +349,44 @@ class CDROM_Interface_Image final : public CDROM_Interface
 
 private:
-	char	device_name[512];
+	void CDAudioCallback(uint16_t requested_frames);
```
CdAudio (lowercase `d`)

johnnovak replied:
`const uint16_t`

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom.h:
```diff
@@ -354,18 +349,44 @@ class CDROM_Interface_Image final : public CDROM_Interface
 private:
-	char	device_name[512];
+	void CDAudioCallback(uint16_t requested_frames);
+	void InitAudio(int device_number);
```
`const int`

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom.h:
```diff
@@ -354,18 +349,44 @@ class CDROM_Interface_Image final : public CDROM_Interface
-	char	device_name[512];
+	void CDAudioCallback(uint16_t requested_frames);
+	void InitAudio(int device_number);
+	bool IsOpen();
```
Can be `bool IsOpen() const;`?

weirddan455 replied:
Fixed

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -83,26 +269,47 @@ bool CDROM_Interface_Ioctl::ReadSectors(PhysPt buffer,
+		close(cdrom_fd);
+	}
+	cdrom_fd = fd;
+	return true;
```
Maybe put a comment there that `Open` also closes the currently open device (so effectively it "re-opens").

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -83,26 +269,47 @@ bool CDROM_Interface_Ioctl::ReadSectors(PhysPt buffer,
+
+bool CDROM_Interface_Ioctl::SetDevice(const char* path, const int cd_number)
+{
+	assert(path != nullptr);
```
minor: you could also assert for a non-blank string, I guess.

weirddan455 replied:
Asserted

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -83,26 +269,47 @@ bool CDROM_Interface_Ioctl::ReadSectors(PhysPt buffer,
-			safe_strncpy(device_name, tmp, 512);
-		else
-			success = false;
+	for (int i = 0; i < num; i++) {
```
minor:
`auto i`
`++i`

weirddan455 replied:
auto'd

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -83,26 +269,47 @@ bool CDROM_Interface_Ioctl::ReadSectors(PhysPt buffer,
+			return true;
+		}
 	}
 
```
Maybe log a warning if we couldn't find it?

weirddan455 replied:
warned

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +320,159 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+
+	std::string name = std::string(MixerChannelNamePrefix) +
+	                   std::to_string(device_number);
+	audio_buffer.resize(AudioBufferBytes / 2);
```
I don't quite like this; can you derive the number of _samples_ instead from the constants and call this `AudioBufferSamples`? Then you don't need to divide by two.

To be clear:

1 sample frame = 2 samples (always stereo)
1 sample = int16_t

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +320,159 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+	                                        frames_available);
+
+	// audio_position must be doubled since a frame is two int16_t's for stereo sound
+#ifdef WORDS_BIGENDIAN
```
Ok, forgot to mention this before... It gets all quite murky and confusing with the samples vs frames business. Some people also use the two interchangeably (not you, luckily 😅), which doesn't help either...

We introduced `AudioFrame` (`include/audio_frame.h`) for this very reason to simplify audio stream handling. I know exactly the distinction between frames and samples, but I managed to still confuse myself sometimes when doing buffer length manipulations, etc.; it's very easy to introduce bugs this way.

We have standardised on the use of `AudioFrames` in all audio rendering code that feeds the mixer. There are a few audio renderers that haven't been converted yet, but that's the ultimate goal. When that goal is accomplished, we will only keep `AddSamples_mfloat` and `AddSamples_sfloat`; every other `AddSample_*` variant will be nuked. That will simply `mixer.cpp` a lot too.

Some example on the use of `AudioFrame`:

#### gus.cpp
```cpp
using pan_scalars_array_t = std::array<AudioFrame, PAN_POSITIONS>;
```
#### gameblaster.cpp
```cpp
std::queue<AudioFrame> fifo = {};
```

####  opl.h
```cpp
std::queue<AudioFrame> fifo = {};
```

####  Misc signatures
```cpp
AudioFrame Process(const AudioFrame &in);
AudioFrame RenderFrame();
bool MaybeRenderFrame(AudioFrame &frame);
void Envelope::Process(const bool is_stereo, AudioFrame &frame)
```

The we can consistently use "frames" in our namings, and everything is crystal clear! For example:

```cpp
	while (frames_remaining) {
		if (float frame = 0.0f; MaybeRenderFrame(frame)) {
			channel->AddSamples_mfloat(1, &frame);
		}
		--frames_remaining;
	}
```

I should have said this earlier... Given our goals stated above, can I ask you to please migrate to using `AudioFrame` and use the term "frame" everywhere consistently? Thanks 😄

weirddan455 replied:
That `AudioFrame` struct only stores floats (well it has a constructor for `int16_t` but it implicitly converts to float).

I'm a little confused how you want me to use this.  I've done a bit of audio coding in the past so I understand the value of wanting to do all your mixing in floats.  However, it's also my understanding you usually want to normalize your floats from -1.0f to 1.0f so I don't know if this cast will cut it.

Like you said earlier, CD audio is different from a SoundBlaster emulator or what have you in that I'm literally just reading `int16_t`'s off a disk.

johnnovak replied:
All good questions, and I'm here to help! Feel free to ask anything about audio if something is unclear.

Yes, using a `[-1.0f, 1.0f]` value range is common in audio processing, but that's just a convention—you can use any range you want, of course.

Currently, the "native" range of the mixer is `[INT16_MIN, INT16_MAX]` internally. You can see evidence of that here; the master compressor I introduced assumes `INT16_MIN` for the [0 dBFS](https://en.wikipedia.org/wiki/DBFS) level (more or less equivalent to "peak amplitude" in common language).
https://github.com/dosbox-staging/dosbox-staging/blob/main/src/hardware/mixer.cpp#L456

So eventually everything has to be normalised to `[INT16_MIN, INT16_MAX]` in our mixer, but you can configure the _expected_ sample value ranges per mixer channels individually with `SetPeakAmplitude()` and `Set0dbScalar()` (expected in the sense of what you pass into the channel with `AddSamples_*`):
https://github.com/dosbox-staging/dosbox-staging/blob/main/src/hardware/mixer.cpp#L840-L845
https://github.com/dosbox-staging/dosbox-staging/blob/main/src/hardware/mixer.cpp#L604-L614

You can read about the whole volume/peak scaling business here:
https://github.com/dosbox-staging/dosbox-staging/blob/main/include/mixer.h#L261-L295

By default, the peak of a mixer channel is `MAX_AUDIO`
https://github.com/dosbox-staging/dosbox-staging/blob/main/include/mixer.h#L291-L295

which is basically `INT16_MAX`
https://github.com/dosbox-staging/dosbox-staging/blob/main/include/mixer.h#L68

So the TL;DR is that yes, you can just cast the raw `int16_t` samples to `float` and simply use `Add_Samples_sfloat` and things will work. That way you will push samples into the mixer channel in our "native" `[INT16_MIN, INT16_MAX]` range. That's a lot of words for confiming "yes, just case to float", but I guess I wanted to explain it to you how it all fits together 😎 

Maybe it's a bit confusing and abstract, but check out how different audio devices set these mixer channel parameters, that should clear things up. However, you only need to cast to floats here.

johnnovak replied:
(And yes, probably I will refactor this all at some point so we use the more common `[-1.0f, 1.0f]` range. There are evolutionary reasons for normalising to the `int16_t` range, but I will rework that when I add 24-bit and 32-bit float audio capture support; then only the `[-1.0f, 1.0f]` internal range will make sense which can be written straight to disk for the 32-bit float capture).

weirddan455 replied:
Interesting.  Yeah, if you get it normalized to -1.0f to 1.0f, you can also feed that directly to SDL (had a quick look and we're currently doing a conversion back to `int16_t` first).  It can accept floating point format but I believe it expects it to be normalized in that manner.

Most modern sound systems can accept float (I know PulseAudio in Linux will) but if not SDL will automatically convert it to what the platform needs.  I vaguely remember reading that Mac OSX **only** accepts floats so if that's true you end up going `int16_t->float->int16_t` then back to float again once SDL does the final conversion internally.

johnnovak replied:
Yeah, maybe 🤷🏻 0.00001% performance drop, I guess 😄 But it will be [-1.0f, 1.0f] anyway for the above outlined reasons in the future.

weirddan455 replied:
Anyway, back to this PR.  I'm still not sure how you want me to use this `AudioFrame` struct.  I can do the conversion from `int16_t` to `float` in my callback function, no problem.  I think that would require two buffers (an `int_16` "read_buffer" to pass to the `ioctl` and then an `output_buffer` with float data to pass to the mixer.  But that's fine if that's what you want.

But then you said you wanted my buffer to be a vector of `AudioFrame`.  Problem is... the mixer doesn't accept that type.  It needs to be a vector of float.  I could `reinterpret_cast` and hope it works out (and it probably will until someone makes changes since the data is just 2 floats) but I think it's bordering on undefined behavior.

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +320,159 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+	                                                (audio_position * 2));
+#else
+	mixer_channel->AddSamples_s16(frames_to_write,
+	                              audio_buffer.data() + (audio_position * 2));
```
Yeah, so `audio_position * 2` would simply become `frame_pos` as your buffer would be a vector (or `std::array`, up to you) of `AudioFrame`s.

weirddan455 replied:
I renamed `audio_position` to `frame_position` but it still has to be multiplied by 2 because the vector is in samples.  See my above comment about `AudioFrame`.  Need some clarification there.

johnnovak replied:
As per my long explanation. You'll have this instead:

```
std::vector<AudioFrame> audio_buffer = {};
```

Then you simply pass yout `int16_t` samples to this contructor:

```
constexpr AudioFrame(const int16_t l, const int16_t r)
```

And push the AudioFrames into the vector. That's it. Make sure to remove every trace of referring to "samples"; the code should only refer to "frames" after this.

weirddan455 replied:
>  And push the AudioFrames into the vector. Add the results to the channel with AddSamples_sfloat. That's it.

`AddSamples_sfloat` won't take a vector of `AudioFrame` is the problem.  I'd be creating a vector that I can't really do anything with.  The add samples function just wants a raw pointer to some floats.

johnnovak replied:
Just do this like every other audio device that uses `AddSamples_sfloat`. Look up the usages, please. `audio_frames` is a `std::vector<AudioFrame>` in this example:

```cpp
channel->AddSamples_sfloat(requested_audio_frames, &audio_frames[0][0]);
```

Again, look at the bigger picture. When everything in the codebase will only push floats to the mixer via `AddSamples_sfloat` and `AddSamples_mfloat`, we can remove that huge templated `AddSamples` junk and introduce two much simpler `AddSamplesStereo` and `AddSamplesMono` functions or something that just take `std::vector<AudioFrame>` directly.

One step at a time...

johnnovak replied:
Okay, I'm kinda busy now, this is surprising. There shouldn't be _that_ much of a performance difference... but then, debug builds are naturally 4-5x slower, or a lot more in some cases. Consider switching to `std::array` perhaps? But probably you'd experience the same relative slowness in debug builds because of the bounds checks or something.

At the end of the day, only the performance of the release build matters.

johnnovak replied:
Btw, only use emplace, that's the potentially more performant version as it skips creating the temporary.

Don't use debug builds for judging the performance of any code, it's meaningless.

weirddan455 replied:
So I was definitely measuring the wrong thing when I said "release builds are mostly the same".  That was wrong.  I did a micro-benchmark of just the loop that was converting 1 second of audio from int to float.

Debug with push_back - 3ms
Debug with emplace_back - 10ms
Release build with either (times are about the same) - 150 **microseconds**

So yeah... almost two orders of magnitude slower in the worst case.  I see your point about not trusting debug performance now.

I think this won't ever be a problem with release builds based on that but I will go ahead and take your suggestion to do the conversion as needed.  If anything, it saves a bit of memory.

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +320,159 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+#endif
+
+	audio_position += frames_to_write;
+	if (audio_position >= frames_in_buffer) {
```
Yeah so this gets confusing... `audio_position` seems to be a "frame position"... but it's not called that.

weirddan455 replied:
renamed to `frame_position`

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +320,159 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+	}
+
+	// Adjust the volume of our mixer channel as defined by the application
+	mixer_channel->SetAppVolume(ctrl.vol[0] / 255.0f, ctrl.vol[1] / 255.0f);
```
Extract `255.0f` to `MaxVolume`.

weirddan455 replied:
extracted

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +320,159 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+	// application
+	const auto left_mapped  = static_cast<LINE_INDEX>(ctrl.out[0]);
+	const auto right_mapped = static_cast<LINE_INDEX>(ctrl.out[1]);
+	mixer_channel->ChangeChannelMap(left_mapped, right_mapped);
```
🥇 for getting this right. I assume you tested it that reversing channels via the mixer works?

weirddan455 replied:
I'm don't know what the mixer command is for reversing.  It's not in the help page.  I just copied that entire function from `cdrom_image.cpp` so I assume it works.  I know the volume setting does by using Quake's volume slider and seeing the debug logs.

johnnovak replied:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/698770/553bfc71-eda0-4efb-be3f-fc4a3fc152f1)

weirddan455 replied:
Not sure how I missed that.  Anyway, my ears really can't tell a difference.  Also, it looks like the mixer code never calls into that function anyway.  That function gets called when a game tries to modify something.  When the user uses the "mixer" command, it just modifies the mixer channel directly without going through the drive.

https://github.com/dosbox-staging/dosbox-staging/blob/40f2a0ac547a53bf4ccf95906a82cb16ec9fed4d/src/hardware/mixer.cpp#L2360-L2361

https://github.com/dosbox-staging/dosbox-staging/blob/40f2a0ac547a53bf4ccf95906a82cb16ec9fed4d/src/hardware/mixer.cpp#L1813-L1828

johnnovak replied:
It will be fine :) If you check out the mixe code in more detail, you'll see those left/right mappings are indices; swapping is accomplished by swapping the 0 and 1 indices.

johnnovak replied:
> Anyway, my ears really can't tell a difference. 

Yes, you won't hear much of a difference unless the audio has some strong stereo content with instruments panned hard left or right. Cymbals and hihats are usually either on the left or right; maybe try listening to that (Quake OST is quite industrial, from memory, with lots of drums).

weirddan455 replied:
I mean it's really hard to tell.  My point was that typing `mixer reverse` never calls into the class I wrote so that test isn't really relevant anyway.  That's all done in the mixer code.  It's a different variable even.  `channel_map` is for when a DOS application wants to change its mapping.  `output_map` is for when the user changes mixer settings (as far as I can tell).

johnnovak replied:
It's a bit more complex, but I don't have time to explain this in detail now... You're doing the same thing as `cdrom_image.cpp`, so it's fine. Moving on 😄

johnnovak commented:
> @johnnovak I think I've addressed all your concerns now. I changed this to do as you said. Good idea on using device ID for the name. That gets rid of the loop checking for available names since we can just assume only 1 channel per device.
> 
> Thanks again for the fast review 😄

My pleasure 😄 You're making great progress and adding new enhancements to Staging at a quick pace! 🎉 

Thanks for addressing the changes, they all look good. I've made another round of comments (I missed some stuff yesterday), but we should be pretty close now!

johnnovak commented on src/dos/program_mount.cpp:
```diff
@@ -462,7 +462,9 @@ void MOUNT::AddMessages() {
-	        "  - '-usecd #' gives direct access to a CD-ROM drive (needed for CD audio).\n"
+	        "  - '-usecd ID' gives direct access to a CD-ROM drive.\n"
+	        "This is needed for CD audio (only supported on Windows and Linux).\n"
+	        "Run 'mount -cd' to find out the list of valid IDs.\n"
```
Nice, thank you 👍🏻

johnnovak replied:
One thing, can you please indent these two new lines with the first line of the bullet point? So just add 4 spaces in front, if I'm counting them right 😄

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -356,45 +363,51 @@ void CDROM_Interface_Ioctl::CdAudioCallback(const uint16_t requested_frames)
+			// For some reason, emplace_back is much slower
+			// in debug build and causes noticible popping
+			// This makes the loop 4ms instead of 10ms on my system.
+			AudioFrame frame(left, right);
```
It's fine, please use emplace and remove this comment. This happens because filling up 1s worth of data takes just a bit too long, which results on starving the audio buffer occasionally in debug builds. You can increase the buffer size in your config to work around this somewhat in debug builds, but it's normal. Debug builds are dog slow with certain settings and can crackle like hell... Also, probably you won't get crackling if you used a standalone audio player program with low cycles settings. Quake with max cycles setting will cause crackles in debug builds, it's not a surprise. Maybe try Monkey Island CD edition with around 3000-5000 cycles. But ultimately it's ok. We could put comments like these into every single file about debug builds 😄

weirddan455 replied:
I've done the on demand float conversations and that seems to have solved the performance issues on debug builds.

johnnovak commented on src/dos/cdrom.h:
```diff
@@ -354,18 +349,44 @@ class CDROM_Interface_Image final : public CDROM_Interface
+	int current_sector                = 0;
+	int sectors_remaining             = 0;
+	bool is_playing                   = false;
+	bool is_paused                    = false;
```
Please reformat.

weirddan455 replied:
It's been squashed and ran through the format script again.

johnnovak commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +324,175 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+	// Output buffer uses C++ vector operations.
+	// Its max capacity will be 1 second so go ahead and reserve that.
+	// Stored in AudioFrame structs.
+	output_buffer.reserve(REDBOOK_PCM_FRAMES_PER_SECOND);
```
Yep, as explained, better not to do this. Only convert as many `int16` samples from the `input_buffer` that the audio callback requested to even out performance spikes. And even then let the `output_buffer` resize itself lazily—the number of requested samples will hover around the same value.

weirddan455 replied:
Reserve has been removed.  Output buffer will be much smaller now (45 samples or so instead of a full second).

weirddan455 commented on src/dos/cdrom_ioctl_linux.cpp:
```diff
@@ -113,4 +324,175 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+	frame_position += frames_to_write;
+	if (frame_position >= output_buffer.size()) {
+		frame_position = 0;
+	}
```
Great explanation.  I did actually do that logging while I was initially writing this and was curious why it was being called so frequently and asking for such a low amount of samples.  Now I know why.

I am a little bit concerned now about the `ioctl` call in the callback.  It was taking 3ms or so to complete (which seems bad if the entire tick is supposed to take 1ms, though the spike only happens once per second) but I haven't noticed any problems in practice.  The `input_buffer` size could be reduced.  It can be anywhere from 1 to 75 redbook frames (75 being 1 second).  I was just using the max to reduce the number of  total syscalls and reducing it probably just means smaller but more frequent latency spikes.

Ultimately though, it's reading from a slow optical drive so I don't know if there's much that can be done short of doing reads on a separate thread.  So maybe it's fine how it is if it doesn't appear to cause any problems?

johnnovak replied:
Finding the optimal number of redbook frames to read is easy enough to do, so I suggest to play around with it a bit (in release builds), and see if you can lower the call latency.

Ultimately, if this works well enough, then I'm happy for you to merge it. Yes, threading  is the best way to go about it, but I don't want to test your patience 😅 ; you've done a stellar job on this feature already! 🚀 🎉 

However, if you're _really_ interested in adding threading (entirely optional!), I'm here to help. Even then, I suggest merging this and adding threading in a secon PR. The patterns are already laid out, so it's a matter of following the `midi_mt32.cpp` or `midi_fluidsynth.cpp` implementations that use `std::thread` and our nice `RWQueue` (in `rwqueue.h`). But more on that if you ask me 😄

johnnovak commented:
This would be a better commit name (to stress this is for Linux only):

Implement CD Audio on Linux using ioctls

weirddan455 commented:
I've renamed the commit.  I also did some quick performance testing in a release build and reduced the maximum number of redbook frames to read from 75 to 50.  This reduces the spikes from 4.1ms average to 2.5ms and slightly lowers the total time spent in ioctls over 10 seconds (38ms down to 37ms).

It's never going to be very fast though without threading.  Just to test, I changed it to read max of 1 redbook frame.  This takes 1ms and increases total time spent in ioctls over 10 seconds to 670ms.  Definitely seeing diminishing returns here.  Although even with this (very sub-optimal) configuration, I did not notice any pops in the audio.

I also created some constants at the top of the file so this value can be tweaked in the future by simply modifying the `InputBufferMaxRedbookFrames` constant.

With that said, I think this is now good to merge if everything looks OK to you.

johnnovak commented:
> also did some quick performance testing in a release build and reduced the maximum number of redbook frames to read from 75 to 50. This reduces the spikes from 4.1ms average to 2.5ms and slightly lowers the total time spent in ioctls over 10 seconds (38ms down to 37ms).

That's an improvement. I get it that the overall time spent when reading a singles frames per call would be way too high, but what if you say read 25 frames at a time instead of 50? If the total time increases even by 30-50% but the spikes are cut in half again, that would make it well worth it.

Perhaps you can try that quickly (if you want), but approving otherwise. Nice work, and thanks for taking my comments on board 👍🏻

johnnovak reviewed (approved):
Great job @weirddan455 🎉

weirddan455 commented:
At 25 frames, it's about 1.7ms average and total time of 50ms.  I'm not sure if this is an improvement or not.  This is where the diminishing returns start.

johnnovak commented:
> At 25 frames, it's about 1.7ms average and total time of 50ms. I'm not sure if this is an improvement or not. This is where the diminishing returns start.

Yeah, I think it's well worth it. Let's do 25 frames then, and then let's merge this. Stutter-free audio is the most important—slightly higher CPU usage, who cares 😄

weirddan455 commented:
>   Yeah, I think it's well worth it. Let's do 25 frames then, and then let's merge this. Stutter-free audio is the most important—slightly higher CPU usage, who cares smile

Right, I wasn't concerned about CPU usage.  This reads twice as often so you hit that potential stutter twice as often was my line of thinking.  But like I said, I don't really have a way to gauge which is better (both sound find on my system).

So changed to 25.
